### PR TITLE
Fix auto guessing when importing CSV

### PR DESCRIPTION
**Pull request by moicalcob**
### Resources

[Clubhouse story](https://app.clubhouse.io/cartoteam/story/159377/import-api-guessing-falls-back-with-wrong-dates)

### Context

When a user tries to import a CSV file with wrong fields the auto guessing feature is not working. The CSV should be imported trying to guess the rest of the columns.

### Changes

- [x]  Import CSV with wrong fields without disabling auto guessing.
- [x]  Add tests

### CODE REVIEW

**Review by Shylpx — CHANGES_REQUESTED** [review]

**Comment by Shylpx** [comment]
Nice job @moicalcob! :smile:, I just added a few tips to improve the readability. :rocket:

File: services/importer/spec/acceptance/csv_spec.rb [hunk 1]
```diff
@@ -377,8 +377,21 @@
...
+    runner = runner_with_fixture('wrong_date_auto_guessing.csv', nil, true)
+    runner.run
+
+    result = runner.results.first
+    age = @user.in_database[%{
+      SELECT *
+      from #{result.schema}.#{result.table_name}
+      }].first.fetch(:age)
+
+    age.class.should eq Integer
```
[the remark this answers is not in the document]
Comment (amiedes) [reply services/importer/spec/acceptance/csv_spec.rb:390]: Indeed I'd favor `expect(age).to be_an(Integer)`

https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/type-matchers

File: services/importer/spec/acceptance/csv_spec.rb [hunk 2]
```diff
@@ -377,8 +377,21 @@
     runner.results.first.success?.should eq true
   end
 
+  it 'auto guessing should work in csv columns if one field is causing troubles' do
```
Comment (amiedes) [inline services/importer/spec/acceptance/csv_spec.rb:380]: I'd try to be more explicit in the test description about what its "causing troubles". Taking a look at the CSV maybe could be: `it 'guesses field types correctly when the CSV contains invalid dates'` or something like that.

File: services/importer/spec/fixtures/wrong_date_auto_guessing.csv [hunk 3]
```diff
@@ -0,0 +1,12 @@
+age,date,name
```
Comment (amiedes) [inline services/importer/spec/fixtures/wrong_date_auto_guessing.csv:1]: I'd try to reduce this file to the minimum text needed to reproduce the issue as that helps keeping clear the intent  of the test case.

I'm not 100% familiar with the ticket but maybe this should be enough?

```csv
age,date
20,2000-01-01
21,2000-1-NA
```

**Review by amiedes — CHANGES_REQUESTED** [review]
Nice catch 💪  . I've left a few comments which should be easy to address, ping me if you have any doubts or want to discuss anything further 🙂 .

**Comment by amiedes** [comment]
In case you're curious about why I'm suggesting this:

> Indeed I'd favor expect(age).to be_an(Integer)

https://github.com/rspec/rspec-expectations/blob/master/Should.md#why-switch-over-from-should-to-expect

File: services/importer/lib/importer/loader.rb [hunk 4]
```diff
@@ -385,6 +387,14 @@ def is_shp?
       def statement_timeout?(error)
         error =~ /canceling statement due to statement timeout/i
       end
+
+      def try_fix_invalid_field(filepath, command_output)
+        line = command_output.split('line')[1].split(':')[0].split(',')[0].delete(' ').to_i - 1
+        column = command_output.split('column')[1].split(':')[0].split(',')[0].delete(' ')
```
[the remark this answers is not in the document]
Comment (moicalcob) [reply services/importer/lib/importer/loader.rb:393]: @Shylpx this is the command output:

```
"ERROR 1: COPY statement failed.\nERROR:  date/time field value out of range: \"2021/02/29 00:00:00\"\nCONTEXT:  COPY importer_244f5bf7196b4d2a8e791bbc59ef891c, line 89, column time: \"2021/02/29 00:00:00\"\n\n"
```

I've been reviewing the proposed changes and I've introduced all of them, but I don't really know how to replace the split functions using regular expressions!

[the remark this answers is not in the document]
Comment (Shylpx) [reply services/importer/lib/importer/loader.rb:393]: @moicalcob You can use something like the following to capture the number of the line:

```
line = /(?<=, line )\d+(?=,)/.match(command_output).to_s.to_i
```
I invite you to create the one for the name of the colum :). Resources:

- Information about regular expressions, incluiding look ahead/look behind: https://www.regular-expressions.info/lookaround.html.
- You can test your regular expresions live with this tool: https://rubular.com/
- And the documentation about the Regexp Ruby class: https://ruby-doc.org/core-2.4.0/Regexp.html

**Review by Shylpx — APPROVED** [review]

**Review by amiedes — APPROVED** [review]

**Pull request merged**